packages/ingest/recursion.js

```diff
@@ -1,84 +1,93 @@
+ * both leading and terminating slashes. List functions will need to be able to handle
+ * or remove leading and terminating slashes accordingly.
+ *
+ * @param {Function} fn - list function
+ * @param {string} currentPath - current path to list
+ * @param {Array<string>} segments - path segments
+ * @param {number} position - current position in the segment list
+ */
+async function recurOnDirectory(fn, currentPath, segments, position) {
+  // check if we have a filter regex segment (e.g. '(dir.*)')
+  const filterExpr = segments[position + 1] || '';
+  const isRegex = isRegexSegment(filterExpr);
 
-    // if directory is found, apply recursion logic
-    if (item.type === 'd' || item.type === 1) {
-      // first we check if the next segment of the provided path
-      // is a regex rule
-      const isRegex = map[position + 1];
+  const path = currentPath.replace(/\/+/g, '/');
+  const contents = fn(path);
+  let files = [];
 
-      // if it is the regex rule, we use the rule to
-      // decide whether to do more recursion
-      if (isRegex) {
-        regexPath = new RegExp(rules[position + 1]);
-      } else {
-        // if it is just a path, we use that segment and its children
-        textPath = rules[position + 1];
+  for (let ctr = 0; ctr < contents.length; ctr += 1) {
+    const item = contents[ctr];
+    if (['-', 0].includes(item.type)) {
+      if (!isRegex || item.name.match(new RegExp(filterExpr))) {
+        // add file to the list if it matches rule
+        files.push(item);
       }
-
-      //  if next segment is regex and matches the rule
-      //  list
-      if (isRegex && item.name.match(regexPath)) {
-        newPath = join(path, item.name);
-      } else if (textPath) {
-        // if it is a regular path, use the path
-        newPath = join(path, textPath);
-      } else {
-        // and ignore all cases that does't match this rule
-        continue; // eslint-disable-line no-continue
+    } else if (['d', 1].includes(item.type)) {
+      // check if dir matches rule
+      if (!isRegex || item.name.match(new RegExp(filterExpr))) {
+        // eslint-disable-next-line no-await-in-loop
+        files = files.concat(await recurOnDirectory(fn, `${path}/${item.name}/`, segments, position + 1));
       }
-
-      // eslint-disable-next-line no-await-in-loop
-      const tmp = await recursion(fn, originalPath, newPath, position + 1);
-      files = files.concat(tmp);
-
-      if (textPath) break;
-    } else if (item.type === '-' || item.type === 0) {
-      // add file to the list
-      files = files.concat(item);
     }
   }
-
   return files;
 }
 
+/**
+ * Recursively list contents of a directory, filtering on provided regex segments.
+ *
+ * @param {Function} fn - list function
+ * @param {string} path - path string which may contain regexes for filtering
+ */
+async function dynamicRecursiveList(fn, path) {
+  const dynamicRegex = /(\([^\)]*\))/g;
```

Comment:
Is this the same regex as for the `isRegexSegment` helper? If so, maybe we make it a `const` for the whole file?

Reply:
Agreed. BTW, pardon my intrusion on your review :-)

Reply:
Done 👍